tools/regression/pkg/builder.go

```diff
@@ -0,0 +1,255 @@
+	BuildJob   BuildJob      // copy of the original build job
+	BuildTime  time.Duration // time it took to build the pkg
+	BinarySize uint64        // size of the binary in bytes
+}
+
+// When the build of a command fails, a BuildError is emitted.
+type BuildError struct {
+	BuildJob BuildJob // copy of the original build job
+	BuildErr string   // error message encountered during build
+}
+
+func (b *BuildError) String() string {
+	return fmt.Sprintf("%v: %v", b.BuildJob.BuildCommand, b.BuildErr)
+}
+
+// Implement error interface over BuildError
+func (b *BuildError) Error() string {
+	return fmt.Sprintf("%v: %v", b.BuildJob.BuildCommand, b.BuildErr)
+}
+
+// BuildDelta provides comparative information about two BuildResults.
+type BuildDelta struct {
+	// size difference of the BuildResults. A negative value means,
+	// that the self object is smaller.
+	deltaSize int64
+	deltaTime time.Duration
+}
+
+// Compare struct other with struct self
+// The BuildDelta will be documented from the perspective
+// of the self object. The goPkgPath of the jobs have to be the same to
+// be comparable; this will be ensured.
+func (b *BuildResult) Delta(o *BuildResult) (BuildDelta, error) {
+	if b.BuildJob.GoPkgPath != o.BuildJob.GoPkgPath {
+		return BuildDelta{}, fmt.Errorf(
+			"cannot compare packages '%v' and %v",
+			b.BuildJob.GoPkgPath,
+			o.BuildJob.GoPkgPath,
+		)
+	}
+
+	if b.BuildJob.Compiler == o.BuildJob.Compiler {
+		return BuildDelta{}, fmt.Errorf(
+			"cannot compare packages with same compiler '%v'",
+			b.BuildJob.Compiler,
+		)
+	}
+
+	return BuildDelta{
+		deltaSize: int64(b.BinarySize) - int64(o.BinarySize),
+		deltaTime: b.BuildTime - o.BuildTime,
+	}, nil
+}
+
+type BuildState int
+
+const (
+	Setup   BuildState = iota // pre-running state. jobs can still be enqueued and config can be changed
+	Running                   // the Builder is running. No new jobs can be queued
+	Stopped                   // the builder was stopped without any errors
+	Error                     // the builder encountered and error
+)
+
+// The Builder struct manages the entire multi-goroutine build process.
+// It starts the worker routines, distributes the BuildJobs, and gathers
+// the BuildResults and BuildErrors.
+type Builder struct {
+	jobQueue    chan BuildJob    // channel for the available build jobs
+	resultQueue chan BuildResult // channel for finished BuildResults
+	errQueue    chan BuildError  // channel for failed build jobs
+	worker      uint             // amount of go work routines
+	state       BuildState       // the current state of the builder
+	jobs        []BuildJob       // list of jobs that will be added to the jobQueue
+	results     []BuildResult    // list of received build results
+	errors      []BuildError     // list of received build errors
+	logger      *log.Logger      // optional, user provided logger
+}
+
+// Generate a new Builder struct with a configuration.
+func NewBuilder(worker uint) (Builder, error) {
```

Comment:
in package build you can use New, `builder.New` instead of `builder.NewBuilder`

Reply:
I named the function the way it is because the builder package contains more than just the Builder struct but many others as well. I wanted to make clear, that this is the constructor for the `Builder` class.